### ISSUE TITLE
AppDir: useState result in infinite loop in client component?

### ISSUE THREAD

**Issue by rnnyrk** [vercel/next.js#50677]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 22.5.0: Mon Apr 24 20:51:50 PDT 2023; root:xnu-8796.121.2~5/RELEASE_X86_64
    Binaries:
      Node: 18.15.0
      npm: 9.5.0
      Yarn: 1.22.19
      pnpm: 7.27.0
    Relevant packages:
      next: 13.4.4
      eslint-config-next: 13.4.4
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.1.3
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App directory (appDir: true)

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/rnnyrk/shopify-tryout/blob/main/src/components/modules/layouts/Navigation.tsx

### To Reproduce

The linked Github repository provides the used code. [It can be viewed live on this url](https://shopify-tryout.vercel.app/en). The issue happens in mobile mode, since it about triggering an mobile menu (see screenshot, also shows the `console.log` of the state loop).

<img width="1613" alt="Screenshot 2023-06-02 at 11 36 16" src="https://github.com/vercel/next.js/assets/4519142/39df718c-ed2c-4de7-b5e4-b3628396d3a2">

### Describe the Bug

My tree looks like
- layout.tsx 
- - RootLayout.tsx `use client`
- - RootContent.tsx `use client`
- - - Navigation.tsx `use client` -> issue happens here

So stripped down this is the code:
```typescript
'use client';
import * as i from 'types';
import clsx from 'clsx';
import { useState } from 'react';

import MenuIcon from 'vectors/menu.svg';
import { CartSheet } from 'modules/cart/CartSheet';

import { LanguageSelect } from './LanguageSelect';
import { MainMenu } from './MainMenu';

export const Navigation = ({ locale }: NavigationProps) => {
  const [isOpen, setOpen] = useState(false);

  console.info({ isOpen });

  return (
    <nav className="w-full h-20 px-4 flex flex-col justify-center">
        <button
          className="appearance-none lg:hidden"
          onClick={() => setOpen(true)}
        >
          <MenuIcon className="w-8 h-8" />
        </button>

      <div
        className={clsx('inset-0 z-20 flex-col',
          {
            flex: isOpen,
            hidden: !isOpen,
          },
        )}
      >
        <MainMenu />

        <div className="flex flex-col items-center lg:flex-row">
          <LanguageSelect
            className="lg:mr-6"
            locale={locale}
          />
          <CartSheet />
        </div>
      </div>
    </nav>
  );
};

type NavigationProps = {
  locale: i.Locale;
};

```

Clicking the button results in an infinite loop of setting the state to (in this case) `true`. But also setting it to the opposite value of the current state results in the same render loop. Also happens on `next 13.4.5-canary.3`.

[This issue ](https://github.com/vercel/next.js/issues/48822) seemed similar, but I can confirm that none of my client components in this tree make use of `async`. The first server component in this three is `LanguageSelect`.

**Edit:**
I also tried the approach with `useReducer` instead of `useState`. Same issue unfortunately.

### Expected Behavior

Only render once after state change.

### Which browser are you using? (if relevant)

Arc / Chromium

### How are you deploying your application? (if relevant)

Vercel

**Comment by icyJoseph** [comment]
Hi,

You have a a children to Navigation, named LanguageSelect which is async. A component part of the `use client` tree is considered to be a client component, and those can't be async. 

You need to use composition: https://nextjs.org/docs/getting-started/react-essentials#recommended-pattern-passing-server-components-to-client-components-as-props

I think you can make [RootContent](https://github.com/rnnyrk/shopify-tryout/blob/main/src/components/modules/layouts/RootContent.tsx) into a server component by removing the `use client` directive, and then use put `children` of `Navigation`:

```tsx
import * as i from 'types';

import { RootLayout } from 'modules/layouts/RootLayout';

import { Hero } from './Hero';
import { Footer } from './Footer';
import { Navigation } from './Navigation';

export const RootContent = ({ children, locale }: RootContentProps) => {
  return (
    <RootLayout>
      <Navigation locale={locale}>

        <LanguageSelect />
      </Navigation>
      <Hero />
      <main>{children}</main>
      <Footer />
    </RootLayout>
  );
};

type RootContentProps = {
  children: React.ReactNode;
  locale: i.Locale;
};
```

And then `Navigation` has to render its children, https://github.com/rnnyrk/shopify-tryout/blob/main/src/components/modules/layouts/Navigation.tsx#L49:

```tsx
 <div className="flex flex-col items-center lg:flex-row">
         {children}
         <CartSheet />
 </div>
```

**Comment by rnnyrk** [comment]
@icyJoseph Thanks for your help. I didn't know about this composition yet. Didn't get the error tho. Updated the latest Typescript and react/types yesterday removing general error of importing async server components. Although it also seems it removes the error where it shouldn't; in client components. 

Adjusting it works.

**Comment by icyJoseph** [comment]
Oh right, I read yesterday that Promise JSX is available in 5.1 right? Uff though luck there. At least it works now 🎉
